File: puzzle7-2.py
import sys
from functools import cmp_to_key

def set_type(hand):
    h = sorted(hand['hand'])
    groups = 0
    jokers = 0
    max_n = 0
    while h:
        c = h.pop(0)
        n = 1
        while h:
            if h[0] == c:
                n += 1
                h.pop(0)
            else:
                break
        if c == 'J':
            jokers = n
        else:
            groups += 1
            if n > max_n:
                max_n = n

    if jokers > 0:
        max_n += jokers
    if jokers == 5:
        groups = 1
    
    if groups == 1:
        hand['type'] = 7 # 5 of a kind
    elif groups == 2:
        if max_n == 4:
            hand['type'] = 6 # 4 of a kind
        else:
            hand['type'] = 5 # full house
    elif groups == 3:
        if max_n == 3:
            hand['type'] = 4 # 3 of a kind
        else:
            hand['type'] = 3 # 2 pair
    elif groups == 4:
        hand['type'] = 2 # 1 pair
    else:
        hand['type'] = 1 # high card


strength = {
    "J": 1,
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "T": 10,
    "Q": 12,
    "K": 13,
    "A": 14
}

def handcmp(hand1, hand2):
    t1 = hand1['type']
    t2 = hand2['type']
    if t1 < t2:
        return -1
    elif t1 > t2:
        return 1
    else:
        for c1, c2 in zip(hand1['hand'], hand2['hand']):
            st1 = strength[c1]
            st2 = strength[c2]
            if st1 < st2:
                return -1
            elif st1 > st2:
                return 1
    assert(0)
    return 0


hands = []
for line in sys.stdin:
    hand, bid = line.rstrip().split()
    hand = dict(hand=hand, bid=int(bid))
    set_type(hand)
    hands.append(hand)
    # print(hand['hand'], hand['type'])

hands.sort(key=cmp_to_key(handcmp))
sum = 0
for i, hand in enumerate(hands):
    sum += (i + 1) * hand['bid']
print(sum)

File: lua/puzzle7-2.lua
require "utils"

local function get_type(hand)
	hand = {unpack(hand)}
	table.sort(hand)

	local groups = {}
	local cur = table.remove(hand)
	assert(cur)
	local n, max, imax = 1, 1, 1
	local len = 0
	local njokers = 0
	while true do
		local c = table.remove(hand)
		if not c or c ~= cur then
			if cur == 'J' then
				njokers = n
			else
				table.insert(groups, n)
				len = len + 1
				if n > max then
					max = n
					imax = len
				end
			end
			if not c then break end
			cur = c
			n = 1
		else
			n = n + 1
		end
	end
	if njokers > 0 then
		if njokers == 5 then
			len = 1
		else
			groups[imax] = groups[imax] + njokers
		end
	end

	if len == 1 then
		return 7 -- 5 of a kind
	elseif len == 2 then
		if groups[1] == 4 or groups[2] == 4 then
			return 6 -- 4 of a kind
		end
		return 5 -- full house
	elseif len == 3 then
		if groups[1] == 2 or groups[2] == 2 then
			return 3 -- 2 pair
		end
		return 4 -- 3 of a kind
	elseif len == 4 then
		return 2 -- 1 pair
	else
		return 1 -- high card
	end
end

local labels = {"J", "2", "3", "4", "5", "6", "7", "8", "9", "T", "Q", "K", "A"}
local strength = {}
for i, v in ipairs(labels) do
	strength[v] = i
end

local function cmp(hand1, hand2)
	if hand1 == hand2 then return false end -- same hand
	if hand1._type == hand2._type then
		local c1, c2 = hand1.cards, hand2.cards
		for i = 1, #c1 do
			local s1, s2 = strength[c1[i]], strength[c2[i]]
			if s1 ~= s2 then return s1 < s2 end
		end
		assert(false) -- duplicates
	else
		return hand1._type < hand2._type
	end
end

local hands = {}
for line in io.lines() do
	local chars, bid = unpack(string.split(line))
	assert(chars, bid)
	local cards = {}
	for i = 1, #chars do
		table.insert(cards, string.sub(chars, i, i))
	end
	hand = {cards=cards, bid=bid + 0, _type=get_type(cards)}
	table.insert(hands, hand)
end

table.sort(hands, cmp)
local sum = 0
for i, hand in ipairs(hands) do
	sum = sum + i * hand.bid
end
print(sum)
